staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go

```diff
@@ -18,34 +18,59 @@ package cache
 
 import (
 	"fmt"
+	"math"
 	"strings"
 	"sync"
 	"testing"
 	"time"
 
 	"k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/clock"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
 	fcache "k8s.io/client-go/tools/cache/testing"
 )
 
+const timeFmt = "15:04:05.999999999"
+
 type testListener struct {
-	lock              sync.RWMutex
-	resyncPeriod      time.Duration
-	expectedItemNames sets.String
-	receivedItemNames []string
-	name              string
+	t                *testing.T
+	name             string
+	clock            clock.PassiveClock
+	resyncPeriod     time.Duration
+	checkTiming      bool // whether handle() checks timing
+	expectedItemKeys sets.String
+	lock             sync.RWMutex
+	whenReceived     map[string]time.Time // maps key to when last received
+}
+
+func (l *testListener) receivedItemKeys() sets.String {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return sets.StringKeySet(l.whenReceived)
+}
+
+func (l *testListener) reset() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.whenReceived = make(map[string]time.Time)
+}
+
+func newTestListener(t *testing.T, name string, clock clock.PassiveClock, resyncPeriod time.Duration, expected ...string) *testListener {
+	return newTimableListener(t, name, clock, resyncPeriod, false, expected...)
 }
 
-func newTestListener(name string, resyncPeriod time.Duration, expected ...string) *testListener {
+func newTimableListener(t *testing.T, name string, clock clock.PassiveClock, resyncPeriod time.Duration, checkTiming bool, expected ...string) *testListener {
```

Comment:
this has enough arguments that it almost wants to be a configuration struct instead.

Consider refactoring a bit, especially since there's a unqualified boolean in the middle, which is not obvious when reading the code.

For instance:

```go
type testConfig struct {
  clock clock.PassiveClock
  expected []string
  t *testing.T
  namePrefix string
}
func (tc *testConfig) listenerAtInterval(interval time.Duration) {
  l := &testListener {
    t: tc.t,
    name: fmt.Sprintf("%s-listener-%s", namePrefix, interval),
    clock: tc.clock,
    resyncPeriod: interval,
    checkTiming: true,
    expectedKeys: sets.NewString(tc.expected...),
 }
 l.reset()
 return l
}

// later
tc := &testConfig{...}
listener1 := tc.listenerAtInterval(0)
listener2 := tc.listenerAtInterval(2*time.Second)
```

Reply:
you can add a second function for unchecked, which then becomes `tc.listenerWithoutTimeChecking(1*time.Second)` on the other tests